compiler/rustc_infer/src/infer/mod.rs

```diff
@@ -334,6 +335,8 @@ pub struct InferCtxt<'tcx> {
     pub intercrate: bool,
 
     next_trait_solver: bool,
+
+    pub fulfilled_obligations: Rc<RefCell<Vec<traits::FulfilledObligation<'tcx>>>>,
```

Comment:
I think the idiomatic thing for the compiler is to use `rustc_data_structures::sync::{Lrc, Lock}` in place of `Rc` and `RefCell`, which abstract over thread-safety.